Generate properties

belka-ew opened the pull request:
Property functions has some additional features. See https://dlang.org/spec/function.html#property-functions for the full list.

belka-ew committed:
Generate properties

lindt commented:
Advantages of @property-usage should be named somehow. Could be in a test case and probably also in the ReadMe.

belka-ew committed:
Add test for type of the generated property.

belka-ew commented:
The most relevant property is imho:
> For the expression typeof(exp) where exp is an @&#8203;property function, the type is the return type of the function, rather than the type of the function.

I added a test at the end for testing if the @&#8203;property has the correct type.

linkrope commented on src/accessors.d:
```diff
@@ -666,3 +666,26 @@ unittest
+    with (new C)
+    {
+        static assert(is(typeof(foo) == int));
+        static assert(!is(typeof(foo) == typeof(fooMethod)));
```
Consider removing this second assertion: it would also pass if you made mistakes in defining `fooMethod`.

belka-ew replied:
I have an idea. I just will append similar static asserts to the existing tests and remove this one. We have tests for const and so on. It would more reliable.

belka-ew committed:
Add tests for type of the generated properties

belka-ew commented:
I added a few new tests in the existing unittest blocks and removed the questionable static assert at the end. @linkrope, @lindt can you take a look again?

duselbaer commented:
Can we merge this PR?

We would like to see a 1.1.0 with at least including the fix for properties of derived classed. Either with or without the feature implemented in this PR.

belka-ew commented:
I can merge it and make a release, I'm just waiting for an OK from @linkrope .

linkrope commented on src/accessors.d:
```diff
@@ -664,5 +678,28 @@ unittest
+        @Read
+        private int foo_;
+
+        final inout(int) fooMethod() inout
```
`fooMethod` is no longer used in the test: remove?

belka-ew replied:
Done

belka-ew committed:
Remove unused test C.fooMethod

belka-ew commented:
Done
